File: socialgym_mobile/lib/utils/weight_unit.dart
/// The unit a weight value (exercise load or bodyweight) is displayed and
/// entered in. All weight is always stored as kilograms — this only governs
/// the client-side display/input boundary.
enum WeightUnit {
  kilograms,
  pounds;

  static const double _kgPerLb = 0.45359237;

  /// The unit conventionally associated with a language, used only until the
  /// person makes an explicit choice in Settings.
  static WeightUnit defaultForLanguageCode(String? languageCode) {
    return languageCode?.toLowerCase().startsWith('en') == true
        ? WeightUnit.pounds
        : WeightUnit.kilograms;
  }

  static WeightUnit fromString(String? value) {
    return WeightUnit.values.firstWhere(
      (e) => e.wireValue.toLowerCase() == value?.toLowerCase(),
      orElse: () => WeightUnit.kilograms,
    );
  }

  /// Backend wire value (`WeightUnit::from_string`/`Display` in
  /// workout/business/src/domain/enums.rs).
  String get wireValue =>
      this == WeightUnit.pounds ? 'Pounds' : 'Kilograms';

  /// Short unit label for display next to a value.
  String get label => this == WeightUnit.pounds ? 'lbs' : 'kg';

  /// Converts a canonical kilogram value to this unit for display.
  double fromKg(double kg) =>
      this == WeightUnit.pounds ? kg / _kgPerLb : kg;

  /// Converts a value entered in this unit back to canonical kilograms for
  /// storage/transmission.
  double toKg(double value) =>
      this == WeightUnit.pounds ? value * _kgPerLb : value;
}

File: socialgym_web/src/commons/library/weightUnit.js
// Weight (exercise load, bodyweight) is always stored as kilograms. These
// helpers convert to/from the person's display unit at the UI boundary only.

const KG_PER_LB = 0.45359237;

export const KILOGRAMS = 'Kilograms';
export const POUNDS = 'Pounds';

// The unit conventionally associated with a language, used only until the
// person makes an explicit choice in Settings.
export const getDefaultWeightUnit = (language) => {
    return language?.toLowerCase().startsWith('en') ? POUNDS : KILOGRAMS;
};

// Resolves the unit weight should be displayed/entered in: the explicit
// `weightUnit` setting if the person picked one, else the language default.
export const resolveWeightUnit = (weightUnit, language) => {
    return weightUnit || getDefaultWeightUnit(language);
};

export const weightUnitLabel = (unit) => (unit === POUNDS ? 'lbs' : 'kg');

export const kgToDisplay = (kg, unit) => (unit === POUNDS ? kg / KG_PER_LB : kg);

export const displayToKg = (value, unit) => (unit === POUNDS ? value * KG_PER_LB : value);
